File: frontend/js/views/wage_garnishment.js
// Wage garnishment generator — CCPA Title III cap (lesser of % of disposable or
// amount above 30× minimum wage), via /calc/wage-garnishment.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderWageGarnishment(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.garnish.h1.title">// WAGE GARNISHMENT</span></h1>
        <p class="muted small" data-i18n="view.garnish.hint.intro">
            The maximum a creditor may garnish from a pay period under the federal CCPA (Title III): the
            lesser of a percentage of disposable earnings (25% for ordinary debt, higher for support) and the
            amount by which disposable earnings exceed 30× the minimum wage. It computes both caps, the
            protected floor, and the garnishment. Drafting aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.garnish.h2.inputs">Garnishment inputs</h2>
            <form id="garnish-form" class="inline-form">
                <label><span data-i18n="view.garnish.label.state">State</span>
                    <input type="text" name="state" value="Delaware" required></label>
                <label><span data-i18n="view.garnish.label.creditor">Creditor</span>
                    <input type="text" name="creditor_name" value=""></label>
                <label><span data-i18n="view.garnish.label.employee">Employee</span>
                    <input type="text" name="employee_name" value=""></label>
                <label><span data-i18n="view.garnish.label.employer">Employer (garnishee)</span>
                    <input type="text" name="employer_name" value=""></label>
                <label><span data-i18n="view.garnish.label.disposable">Disposable earnings ($)</span>
                    <input type="number" step="10" min="0" name="disposable_earnings_usd" value="600" required></label>
                <label><span data-i18n="view.garnish.label.minwage">Minimum wage ($/hr)</span>
                    <input type="number" step="0.25" min="0" name="min_wage_usd" value="7.25"></label>
                <label><span data-i18n="view.garnish.label.pct">Cap on disposable (%)</span>
                    <input type="number" step="1" min="0" max="100" name="cap_pct" value="25"></label>
                <label><span data-i18n="view.garnish.label.mult">Min-wage multiplier</span>
                    <input type="number" step="1" min="0" name="min_wage_multiplier" value="30"></label>
                <label><span data-i18n="view.garnish.label.period">Pay period</span>
                    <input type="text" name="pay_period" value="weekly"></label>
                <label><span data-i18n="view.garnish.label.date">Date</span>
                    <input type="date" name="date" value="2026-07-01" required></label>
                <label><span data-i18n="view.garnish.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.garnish.ph.statute'))}"></label>
            </form>
        </div>
        <div id="garnish-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#garnish-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            creditor_name: (fd.get('creditor_name') || '').trim(),
            employee_name: (fd.get('employee_name') || '').trim(),
            employer_name: (fd.get('employer_name') || '').trim(),
            disposable_earnings_usd: Number(fd.get('disposable_earnings_usd')) || 0,
            min_wage_usd: Number(fd.get('min_wage_usd')) || 0,
            cap_pct: Number(fd.get('cap_pct')) || 0,
            min_wage_multiplier: Number(fd.get('min_wage_multiplier')) || 0,
            pay_period: (fd.get('pay_period') || '').trim(),
            date: fd.get('date'),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcWageGarnishment(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.garnish.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const bindKey = {
        percentage: 'view.garnish.bind.percentage',
        floor: 'view.garnish.bind.floor',
        none: 'view.garnish.bind.none',
    }[doc.binding_cap] || 'view.garnish.bind.none';
    const el = mount.querySelector('#garnish-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card neg"><div class="label" data-i18n="view.garnish.card.garnishable">Garnishable</div>
                    <div class="value">${money(doc.garnishable_usd)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.garnish.card.keeps">Employee keeps</div>
                    <div class="value">${money(doc.employee_keeps_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.garnish.card.pctcap">25% cap</div>
                    <div class="value">${money(doc.percentage_cap_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.garnish.card.floor">Protected floor</div>
                    <div class="value">${money(doc.protected_floor_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.garnish.card.binding">Binding cap</div>
                    <div class="value" data-i18n="${bindKey}">${esc(doc.binding_cap)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="garnish-copy" type="button" data-i18n="view.garnish.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="garnish-download" type="button" data-i18n="view.garnish.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#garnish-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.garnish.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.garnish.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#garnish-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'wage-garnishment.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/wage_garnishment.rs
//! Wage garnishment calculation — the maximum amount a creditor may garnish from
//! a pay period under the federal Consumer Credit Protection Act (Title III). The
//! garnishable amount is the lesser of (a) a percentage of disposable earnings
//! (25% for ordinary debt; higher for support) and (b) the amount by which
//! disposable earnings exceed 30× the federal minimum wage. This computes both
//! caps, the protected floor, and the resulting garnishment. Distinct from
//! take-home-paycheck, which computes net pay, not garnishment limits. Drafting
//! aid, not legal advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct GarnishmentInput {
    pub creditor_name: String,
    pub employee_name: String,
    pub employer_name: String,
    /// Disposable earnings for the pay period (after legally required deductions).
    pub disposable_earnings_usd: f64,
    /// Applicable federal (or higher state) minimum wage, per hour.
    #[serde(default = "default_min_wage")]
    pub min_wage_usd: f64,
    /// Percentage cap on disposable earnings (25% ordinary; 50–65% support).
    #[serde(default = "default_pct")]
    pub cap_pct: f64,
    /// Minimum-wage multiplier for the protected floor (30× weekly under CCPA).
    #[serde(default = "default_multiplier")]
    pub min_wage_multiplier: f64,
    pub pay_period: String,
    pub date: String,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

fn default_min_wage() -> f64 {
    7.25
}

fn default_pct() -> f64 {
    25.0
}

fn default_multiplier() -> f64 {
    30.0
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct WageGarnishment {
    pub title: String,
    /// 30× minimum wage — earnings below this are fully protected.
    pub protected_floor_usd: f64,
    /// cap_pct × disposable earnings.
    pub percentage_cap_usd: f64,
    /// Disposable earnings above the protected floor (≥ 0).
    pub above_floor_usd: f64,
    /// The lesser of the percentage cap and the amount above the floor.
    pub garnishable_usd: f64,
    /// Disposable earnings the employee keeps.
    pub employee_keeps_usd: f64,
    /// Which cap bound the result: "percentage", "floor", or "none".
    pub binding_cap: String,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &GarnishmentInput) -> WageGarnishment {
    let floor = cents(i.min_wage_multiplier * i.min_wage_usd);
    let pct_cap = cents(i.disposable_earnings_usd * i.cap_pct / 100.0);
    let above_floor = (i.disposable_earnings_usd - floor).max(0.0);
    let garnishable = cents(pct_cap.min(above_floor));
    let keeps = cents(i.disposable_earnings_usd - garnishable);

    let binding = if garnishable <= 0.0 {
        "none"
    } else if pct_cap <= above_floor {
        "percentage"
    } else {
        "floor"
    };

    let binding_desc = match binding {
        "percentage" => format!("the {:.1}% cap on disposable earnings", i.cap_pct),
        "floor" => format!("the amount above 30× the minimum wage ({})", money(floor)),
        _ => "full protection (earnings at or below the protected floor)".to_string(),
    };

    let calc_body = format!(
        "Disposable earnings are {}. The protected floor is {} ({}× minimum wage of {}). The {:.1}% cap is {}; the amount above the floor is {}. The garnishable amount is the lesser, {}, bound by {}. The employee keeps {}.",
        money(i.disposable_earnings_usd),
        money(floor),
        i.min_wage_multiplier,
        money(i.min_wage_usd),
        i.cap_pct,
        money(pct_cap),
        money(cents(above_floor)),
        money(garnishable),
        binding_desc,
        money(keeps)
    );

    let citation = i.statute_citation.trim();
    let governing = if citation.is_empty() {
        format!("Garnishment is limited by 15 U.S.C. § 1673 and the laws of the State of {}.", i.state)
    } else {
        format!("Garnishment is limited by 15 U.S.C. § 1673 and the laws of the State of {} ({}).", i.state, citation)
    };

    let clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!(
                "Creditor: {}\nEmployee: {}\nEmployer (garnishee): {}\nPay period: {}\nDate: {}",
                i.creditor_name, i.employee_name, i.employer_name, i.pay_period, i.date
            ),
        },
        DocClause {
            heading: "1. Disposable Earnings".into(),
            body: format!(
                "Disposable earnings for the pay period — gross pay less legally required deductions — are {}.",
                money(i.disposable_earnings_usd)
            ),
        },
        DocClause { heading: "2. Garnishment Limit".into(), body: calc_body },
        DocClause {
            heading: "3. Withholding".into(),
            body: format!(
                "The employer shall withhold {} from this pay period and remit it toward the judgment, leaving the employee {}. No greater amount may be withheld for ordinary debt without further order.",
                money(garnishable), money(keeps)
            ),
        },
        DocClause { heading: "4. Governing Law".into(), body: governing },
        DocClause {
            heading: "Certification".into(),
            body: format!(
                "Employer (garnishee): ____________________  Date: __________\n{}",
                i.employer_name
            ),
        },
    ];

    WageGarnishment {
        title: "Wage Garnishment Calculation".into(),
        protected_floor_usd: floor,
        percentage_cap_usd: pct_cap,
        above_floor_usd: cents(above_floor),
        garnishable_usd: garnishable,
        employee_keeps_usd: keeps,
        binding_cap: binding.to_string(),
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> GarnishmentInput {
        GarnishmentInput {
            creditor_name: "Collections LLC".into(),
            employee_name: "Pat Employee".into(),
            employer_name: "Acme Co".into(),
            disposable_earnings_usd: 600.0,
            min_wage_usd: 7.25,
            cap_pct: 25.0,
            min_wage_multiplier: 30.0,
            pay_period: "weekly".into(),
            date: "2026-07-01".into(),
            state: "Delaware".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn percentage_cap_binds() {
        let d = generate(&base());
        assert!(close(d.protected_floor_usd, 217.50));
        assert!(close(d.percentage_cap_usd, 150.0));
        assert!(close(d.above_floor_usd, 382.50));
        assert!(close(d.garnishable_usd, 150.0));
        assert!(close(d.employee_keeps_usd, 450.0));
        assert_eq!(d.binding_cap, "percentage");
    }

    #[test]
    fn floor_cap_binds_for_low_income() {
        let d = generate(&GarnishmentInput { disposable_earnings_usd: 250.0, ..base() });
        // 25% = 62.50; above floor = 32.50 → lesser is 32.50.
        assert!(close(d.garnishable_usd, 32.50));
        assert_eq!(d.binding_cap, "floor");
    }

    #[test]
    fn fully_protected_below_floor() {
        let d = generate(&GarnishmentInput { disposable_earnings_usd: 200.0, ..base() });
        assert!(close(d.above_floor_usd, 0.0));
        assert!(close(d.garnishable_usd, 0.0));
        assert_eq!(d.binding_cap, "none");
        assert!(close(d.employee_keeps_usd, 200.0));
    }

    #[test]
    fn support_higher_percentage() {
        // Child support can reach 50–65%; here 50% of 600 = 300, still below above-floor 382.50.
        let d = generate(&GarnishmentInput { cap_pct: 50.0, ..base() });
        assert!(close(d.garnishable_usd, 300.0));
        assert_eq!(d.binding_cap, "percentage");
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&GarnishmentInput { statute_citation: "Del. Code tit. 10 § 4913".into(), ..base() });
        assert_eq!(d.statutory_citation, "Del. Code tit. 10 § 4913");
        assert!(d.clauses.iter().any(|c| c.body.contains("Del. Code tit. 10 § 4913")));
    }
}
